Slider: Add support for decimal values

samsch opened the pull request:
PR content partially copied from: https://github.com/grafana/grafana/pull/113016

Changelog: Fix - UI: Slider now supports decimal input and clamps on blur.

### Problem
Note that when using Grafana's UI Slider component and setting its min/max/step properties to decimal values, the slider will increase/decrease by decimal values, but the input itself will be set to 0 through errors or not handled properly.

**### Solution**
Modified the Slider component to:
- Allow typing literal partial-decimal values (e.g. "1.", "0.") rather than requiring coercion.
- Support for decimal input values while typing
- Correctly clamp to min/max and step bounds on blur
- Tests updated for new behavior around decimals.

**### Testing**

1. Manual testing:
Type decimal values such as "0.5", "1.7" in the input
- Check partial inputs like "1." aren't immediately coerced
- Ensure values are clamped to min/max/step in the blur.

2. Added/Updated unit tests
- Test decimal input behavior
- Test min/max boundary enforcement
- Test partial decimal input states
- Test integer behavior
- Test controlled behavior

Fixes https://github.com/grafana/grafana/issues/107350

**Special notes for your reviewer:**

This includes handling for floating point errors to display the values in a friendly way to users. If a floating point error value (e.g., 2.0000000000002) shows up, that's a bug!

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

HarshadaGawas05 committed:
UI: Fix Slider component to handle decimal inputs correctly

samsch committed:
format code and run eslint fix

samsch committed:
fix Slider to have:
- "single" source of truth for state
- state synchronization for controlled values
- clamp values to step
- disallow decimal values in input when min+step are integers
- tests for the new functionality
- design decision included in docs
- behavior notes in docs

joshhunt reviewed (commented):
Was a bit confused testing this, as when I tried to type in a decimal it never actually entered - typing `1.5` would just type `15`. Turns out I was missing setting the `step` property! I filled that in and it all worked.

I'm not sure if this is something we want to think about alternate behaviour? I think it's probably _fine_ as is, but is disorientating to type on the keyboard and not have it reflected (even if invalid).

Given we already clamp to max/min, and to the nearest step, maybe we can just allow the user to always type in a decimal and just round it to the nearest step?

samsch committed:
allow non-numeric characters all the time
always parse decimal numbers, stripping non-numerics
integer coercion is implicitly handled in clamping

samsch committed:
Merge branch 'main' into fix-slider-decimal-input

joshhunt reviewed (approved):
👍 nice stuff!!!!
